File: sdks/py/src/loxa/metrics.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from datetime import timedelta
from threading import Lock
from typing import Any, Iterable


DEFAULT_HISTOGRAM_BUCKETS: tuple[float, ...] = (
    0.001,
    0.0025,
    0.005,
    0.01,
    0.025,
    0.05,
    0.1,
    0.25,
    0.5,
    1.0,
    2.5,
    5.0,
    10.0,
)


def _escape_label_value(value: Any) -> str:
    text = str(value)
    return text.replace("\\", "\\\\").replace("\n", "\\n").replace('"', '\\"')


def _format_labels(labels: dict[str, Any]) -> str:
    if not labels:
        return ""
    parts = [f'{key}="{_escape_label_value(value)}"' for key, value in sorted(labels.items())]
    return "{" + ",".join(parts) + "}"


@dataclass(slots=True)
class _Histogram:
    buckets: tuple[float, ...]
    counts: dict[float, int] = field(default_factory=dict)
    total: int = 0
    sum: float = 0.0

    def __post_init__(self) -> None:
        for bucket in self.buckets:
            self.counts.setdefault(bucket, 0)

    def observe(self, value: float) -> None:
        self.total += 1
        self.sum += value
        for bucket in self.buckets:
            if value <= bucket:
                self.counts[bucket] += 1


class MetricsCollector:
    """Thread-safe Prometheus-style metrics collector for the Python SDK."""

    def __init__(
        self,
        namespace: str = "loxa_sdk",
        *,
        buffer_capacity: int = 0,
        histogram_buckets: Iterable[float] = DEFAULT_HISTOGRAM_BUCKETS,
    ) -> None:
        self.namespace = namespace or "loxa_sdk"
        self._lock = Lock()
        self._counters: dict[str, int] = defaultdict(int)
        self._counter_vecs: dict[str, dict[str, int]] = defaultdict(lambda: defaultdict(int))
        self._gauges: dict[str, float] = {
            "buffer_size": 0.0,
            "buffer_capacity": float(max(0, buffer_capacity)),
        }
        self._emit_duration = _Histogram(tuple(sorted(set(float(b) for b in histogram_buckets))))

    def on_event_created(self) -> None:
        with self._lock:
            self._counters["events_created_total"] += 1

    def on_event_finished(self) -> None:
        with self._lock:
            self._counters["events_finished_total"] += 1

    def on_event_emitted(self, success: bool = True) -> None:
        with self._lock:
            key = "success" if success else "failure"
            self._counter_vecs["events_emitted_total"][key] += 1

    def on_event_dropped(self, reason: str) -> None:
        with self._lock:
            self._counter_vecs["events_dropped_total"][reason] += 1

    def on_retry(self, attempt: int) -> None:
        with self._lock:
            self._counter_vecs["retry_total"][str(max(1, int(attempt)))] += 1

    def on_backpressure(self) -> None:
        with self._lock:
            self._counters["backpressure_total"] += 1

    def observe_emit_duration(self, duration: float | timedelta) -> None:
        seconds = duration.total_seconds() if isinstance(duration, timedelta) else float(duration)
        seconds = max(0.0, seconds)
        with self._lock:
            self._emit_duration.observe(seconds)

    def set_buffer_size(self, size: int) -> None:
        with self._lock:
            self._gauges["buffer_size"] = float(max(0, size))

    def set_buffer_capacity(self, size: int) -> None:
        with self._lock:
            self._gauges["buffer_capacity"] = float(max(0, size))

    def snapshot(self) -> dict[str, Any]:
        with self._lock:
            return {
                "counters": dict(self._counters),
                "counter_vecs": {key: dict(value) for key, value in self._counter_vecs.items()},
                "gauges": dict(self._gauges),
                "emit_duration": {
                    "buckets": dict(self._emit_duration.counts),
                    "count": self._emit_duration.total,
                    "sum": self._emit_duration.sum,
                },
            }

    def render_prometheus(self) -> str:
        with self._lock:
            lines: list[str] = []
            self._append_counter(lines, "events_created_total", "Total number of events created via StartEvent")
            self._append_counter(lines, "events_finished_total", "Total number of events finished via Finish/FinishError")
            self._append_counter_vec(
                lines,
                "events_emitted_total",
                "Total number of events emitted to sink",
            )
            self._append_counter_vec(
                lines,
                "events_dropped_total",
                "Total number of events dropped",
            )
            self._append_counter_vec(
                lines,
                "retry_total",
                "Total number of retry attempts",
            )
            self._append_counter(lines, "backpressure_total", "Total number of backpressure events")
            self._append_gauge(lines, "buffer_size", "Current number of events in buffer")
            self._append_gauge(lines, "buffer_capacity", "Maximum buffer size")
            self._append_histogram(lines, "emit_duration_seconds", "Duration of Emit operations in seconds")
            return "\n".join(lines) + "\n"

    def _metric_name(self, name: str) -> str:
        return f"{self.namespace}_{name}"

    def _append_counter(self, lines: list[str], name: str, help_text: str) -> None:
        metric = self._metric_name(name)
        lines.append(f"# HELP {metric} {help_text}")
        lines.append(f"# TYPE {metric} counter")
        lines.append(f"{metric} {self._counters.get(name, 0)}")

    def _append_counter_vec(self, lines: list[str], name: str, help_text: str) -> None:
        metric = self._metric_name(name)
        lines.append(f"# HELP {metric} {help_text}")
        lines.append(f"# TYPE {metric} counter")
        for label, value in sorted(self._counter_vecs.get(name, {}).items()):
            lines.append(f"{metric}{_format_labels({'status' if name == 'events_emitted_total' else 'reason' if name == 'events_dropped_total' else 'attempt': label})} {value}")

    def _append_gauge(self, lines: list[str], name: str, help_text: str) -> None:
        metric = self._metric_name(name)
        lines.append(f"# HELP {metric} {help_text}")
        lines.append(f"# TYPE {metric} gauge")
        lines.append(f"{metric} {self._gauges.get(name, 0.0)}")

    def _append_histogram(self, lines: list[str], name: str, help_text: str) -> None:
        metric = self._metric_name(name)
        lines.append(f"# HELP {metric} {help_text}")
        lines.append(f"# TYPE {metric} histogram")
        cumulative = 0
        for bucket in self._emit_duration.buckets:
            cumulative = self._emit_duration.counts.get(bucket, cumulative)
            lines.append(f'{metric}_bucket{{le="{bucket:g}"}} {cumulative}')
        lines.append(f'{metric}_bucket{{le="+Inf"}} {self._emit_duration.total}')
        lines.append(f"{metric}_sum {self._emit_duration.sum}")
        lines.append(f"{metric}_count {self._emit_duration.total}")


def NewMetricsCollector(namespace: str = "loxa_sdk", *, buffer_capacity: int = 0) -> MetricsCollector:
    return MetricsCollector(namespace=namespace, buffer_capacity=buffer_capacity)


def RenderPrometheus(metrics: MetricsCollector) -> str:
    return metrics.render_prometheus()


MetricsSnapshot = dict[str, Any]
PrometheusMetrics = MetricsCollector


__all__ = [
    "DEFAULT_HISTOGRAM_BUCKETS",
    "MetricsCollector",
    "MetricsSnapshot",
    "NewMetricsCollector",
    "PrometheusMetrics",
    "RenderPrometheus",
]

File: sdks/go/src/core/metrics.go
package core

import (
	"net/http"
	"strconv"
	"sync/atomic"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

// MetricsCollector implements Prometheus metrics for the SDK.
// Requirements: 49.1, 49.2, 49.3, 49.4, 49.5, 49.6, 49.7, 49.8, 49.9, 49.10
type MetricsCollector struct {
	// Counters
	eventsCreated  prometheus.Counter
	eventsFinished prometheus.Counter
	eventsEmitted  *prometheus.CounterVec
	eventsDropped  *prometheus.CounterVec
	retryTotal     *prometheus.CounterVec
	backpressure   prometheus.Counter

	// Gauges
	bufferSize     prometheus.Gauge
	bufferCapacity prometheus.Gauge

	// Histograms
	emitDuration prometheus.Histogram

	// Registry
	registry *prometheus.Registry

	// Internal state for buffer tracking
	currentBufferSize atomic.Int64
	maxBufferSize     int64
}

// NewMetricsCollector creates a new Prometheus metrics collector for the SDK.
// Requirements: 49.10
func NewMetricsCollector(namespace string, maxBufferSize int) *MetricsCollector {
	if namespace == "" {
		namespace = "loxa_sdk"
	}

	registry := prometheus.NewRegistry()

	mc := &MetricsCollector{
		registry:      registry,
		maxBufferSize: int64(maxBufferSize),
	}

	// Counter: events_created_total
	// Requirement: 49.1
	mc.eventsCreated = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "events_created_total",
		Help:      "Total number of events created via StartEvent",
	})

	// Counter: events_finished_total
	// Requirement: 49.2
	mc.eventsFinished = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "events_finished_total",
		Help:      "Total number of events finished via Finish/FinishError",
	})

	// Counter: events_emitted_total with status label
	// Requirement: 49.3
	mc.eventsEmitted = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "events_emitted_total",
			Help:      "Total number of events emitted to collector",
		},
		[]string{"status"},
	)

	// Counter: events_dropped_total with reason label
	// Requirement: 49.4
	mc.eventsDropped = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "events_dropped_total",
			Help:      "Total number of events dropped",
		},
		[]string{"reason"},
	)

	// Histogram: emit_duration_seconds
	// Requirement: 49.5
	mc.emitDuration = prometheus.NewHistogram(prometheus.HistogramOpts{
		Namespace: namespace,
		Name:      "emit_duration_seconds",
		Help:      "Duration of Emit operations in seconds",
		Buckets:   prometheus.DefBuckets,
	})

	// Gauge: buffer_size
	// Requirement: 49.6
	mc.bufferSize = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "buffer_size",
		Help:      "Current number of events in buffer",
	})

	// Gauge: buffer_capacity
	// Requirement: 49.7
	mc.bufferCapacity = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "buffer_capacity",
		Help:      "Maximum buffer size",
	})
	mc.bufferCapacity.Set(float64(maxBufferSize))

	// Counter: retry_total with attempt label
	// Requirement: 49.8
	mc.retryTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "retry_total",
			Help:      "Total number of retry attempts",
		},
		[]string{"attempt"},
	)

	// Counter: backpressure_total
	// Requirement: 49.9
	mc.backpressure = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "backpressure_total",
		Help:      "Total number of backpressure events (429/503 responses)",
	})

	// Register all metrics
	registry.MustRegister(
		mc.eventsCreated,
		mc.eventsFinished,
		mc.eventsEmitted,
		mc.eventsDropped,
		mc.emitDuration,
		mc.bufferSize,
		mc.bufferCapacity,
		mc.retryTotal,
		mc.backpressure,
	)

	return mc
}

// OnEventCreated increments the events_created_total counter.
// Requirement: 49.1
func (mc *MetricsCollector) OnEventCreated() {
	if mc != nil && mc.eventsCreated != nil {
		mc.eventsCreated.Inc()
		mc.incrementBufferSize()
	}
}

// OnEventFinished increments the events_finished_total counter.
// Requirement: 49.2
func (mc *MetricsCollector) OnEventFinished() {
	if mc != nil && mc.eventsFinished != nil {
		mc.eventsFinished.Inc()
	}
}

// OnEventEmitted increments the events_emitted_total counter with status label.
// Requirement: 49.3
func (mc *MetricsCollector) OnEventEmitted(success bool) {
	if mc != nil && mc.eventsEmitted != nil {
		status := "success"
		if !success {
			status = "failure"
		}
		mc.eventsEmitted.WithLabelValues(status).Inc()
		mc.decrementBufferSize()
	}
}

// OnEventDropped increments the events_dropped_total counter with reason label.
// Requirement: 49.4
func (mc *MetricsCollector) OnEventDropped(reason string) {
	if mc != nil && mc.eventsDropped != nil {
		mc.eventsDropped.WithLabelValues(reason).Inc()
		mc.decrementBufferSize()
	}
}

// ObserveEmitDuration records the duration of an Emit operation.
// Requirement: 49.5
func (mc *MetricsCollector) ObserveEmitDuration(duration time.Duration) {
	if mc != nil && mc.emitDuration != nil {
		mc.emitDuration.Observe(duration.Seconds())
	}
}

// OnRetry increments the retry_total counter with attempt label.
// Requirement: 49.8
func (mc *MetricsCollector) OnRetry(attempt int) {
	if mc != nil && mc.retryTotal != nil {
		mc.retryTotal.WithLabelValues(strconv.Itoa(attempt)).Inc()
	}
}

// OnBackpressure increments the backpressure_total counter.
// Requirement: 49.9
func (mc *MetricsCollector) OnBackpressure() {
	if mc != nil && mc.backpressure != nil {
		mc.backpressure.Inc()
	}
}

// incrementBufferSize increments the buffer size gauge.
// Requirement: 49.6
func (mc *MetricsCollector) incrementBufferSize() {
	if mc != nil && mc.bufferSize != nil {
		newSize := mc.currentBufferSize.Add(1)
		mc.bufferSize.Set(float64(newSize))
	}
}

// decrementBufferSize decrements the buffer size gauge.
// Requirement: 49.6
func (mc *MetricsCollector) decrementBufferSize() {
	if mc != nil && mc.bufferSize != nil {
		newSize := mc.currentBufferSize.Add(-1)
		if newSize < 0 {
			newSize = 0
			mc.currentBufferSize.Store(0)
		}
		mc.bufferSize.Set(float64(newSize))
	}
}

// SetBufferSize sets the current buffer size directly.
// Requirement: 49.6
func (mc *MetricsCollector) SetBufferSize(size int) {
	if mc != nil && mc.bufferSize != nil {
		mc.currentBufferSize.Store(int64(size))
		mc.bufferSize.Set(float64(size))
	}
}

// Handler returns an HTTP handler for Prometheus metrics endpoint.
// Requirement: 49.10
func (mc *MetricsCollector) Handler() http.Handler {
	if mc == nil || mc.registry == nil {
		return http.NotFoundHandler()
	}
	return promhttp.HandlerFor(mc.registry, promhttp.HandlerOpts{})
}

// Registry returns the Prometheus registry for custom registration.
// Requirement: 49.11
func (mc *MetricsCollector) Registry() *prometheus.Registry {
	if mc == nil {
		return nil
	}
	return mc.registry
}

// PrometheusStatsHandler wraps MetricsCollector to implement StatsHandler interface.
// Requirements: 34.3, 34.4
type PrometheusStatsHandler struct {
	metrics *MetricsCollector
}

// NewPrometheusStatsHandler creates a new StatsHandler backed by Prometheus metrics.
func NewPrometheusStatsHandler(namespace string, maxBufferSize int) *PrometheusStatsHandler {
	return &PrometheusStatsHandler{
		metrics: NewMetricsCollector(namespace, maxBufferSize),
	}
}

// OnEmit is called when an event is successfully emitted.
// Requirement: 34.3
func (h *PrometheusStatsHandler) OnEmit(ev *Event) {
	if h != nil && h.metrics != nil {
		h.metrics.OnEventEmitted(true)
	}
}

// OnDrop is called when an event is dropped.
// Requirement: 34.4
func (h *PrometheusStatsHandler) OnDrop(reason string) {
	if h != nil && h.metrics != nil {
		h.metrics.OnEventDropped(reason)
	}
}

// OnError is called when an error occurs during asynchronous operations.
// Requirement: 34.3, 34.4
func (h *PrometheusStatsHandler) OnError(err error) {
	_ = err
}

// OnDeliveryFailed records an explicit delivery failure without counting it as a success emit.
func (h *PrometheusStatsHandler) OnDeliveryFailed(_ *Event, _ error) {
	if h != nil && h.metrics != nil {
		h.metrics.OnEventEmitted(false)
	}
}

// OnEventCreated increments the events_created_total counter.
func (h *PrometheusStatsHandler) OnEventCreated() {
	if h != nil && h.metrics != nil {
		h.metrics.OnEventCreated()
	}
}

// OnEventFinished increments the events_finished_total counter.
func (h *PrometheusStatsHandler) OnEventFinished() {
	if h != nil && h.metrics != nil {
		h.metrics.OnEventFinished()
	}
}

// ObserveEmitDuration records the duration of an emit operation.
func (h *PrometheusStatsHandler) ObserveEmitDuration(d time.Duration) {
	if h != nil && h.metrics != nil {
		h.metrics.ObserveEmitDuration(d)
	}
}

// Metrics returns the underlying MetricsCollector for direct access.
func (h *PrometheusStatsHandler) Metrics() *MetricsCollector {
	if h == nil {
		return nil
	}
	return h.metrics
}

// Handler returns an HTTP handler for the Prometheus metrics endpoint.
func (h *PrometheusStatsHandler) Handler() http.Handler {
	if h == nil || h.metrics == nil {
		return http.NotFoundHandler()
	}
	return h.metrics.Handler()
}
